docs/architecture/deployment-execution/import.md

```diff
@@ -0,0 +1,184 @@
+   `Iₚ`. This is intended to produce the smallest valid set of inputs necessary
+   to avoid diffs. This does not use a fixed-point algorithm because there is no
+   guarantee that the values copied from `Sₐ` are in fact valid (state and
+   inputs with the same property paths may have different types and validation
+   rules) and there is no guarantee that such an algorithm would terminate (
+   bridged Terraform providers, for example, have had bugs that cause persistent
+   diffs, which can only be worked around with `ignoreChanges`).
+
+If all of these steps succeed, the user is left with a definition for `R` in
+their state. The Pulumi CLI then passes the inputs `Iₚ` stored in the state to
+the import code generator. The import code generator converts the values present
+in `Iₚ` into an equivalent [PCL](pcl) representation of `R`'s desired state,
+then passes the PCL to a language-specific code generator to emit a
+representation of `R`'s desired state in the language used by the destination
+stack's Pulumi program. The user can then copy the generated definition into
+their Pulumi program.
+
+```{mermaid}
+:caption: Importing a resource using the `pulumi import` CLI
+:zoom:
+
+sequenceDiagram
+    participant PI as pulumi import
+    participant CV as PCL converter
+    participant LH as Language host
+    box Engine
+        participant ID as Import driver
+        participant SE as Step executor
+    end
+    participant P as Provider
+
+    PI->>+ID: Import(specs)
+    ID->>+SE: ImportStep(inputs, options)
+    SE->>+P: ReadRequest(type, id)
+    P->>-SE: ReadResponse(current inputs, current state)
+    SE->>+P: CheckRequest(type, inputs, current inputs)
+    P->>-SE: CheckResponse(inputs', failures)
+    SE->>+P: DiffRequest(type, inputs', current state, options)
+    P->>-SE: DiffResponse(diff)
+    SE->>-ID: Done(current state)
+    ID->>-PI: Done(current state)
+
+    PI->>+CV: Convert(current state)
+    CV->>-PI: Done(PCL definitions)
+
+    PI->>+LH: GenerateProgram(PCL definitions)
+    LH->>-PI: Done(generated resource definitions)
+```
+
+### Limitations and challenges
+
+The primary challenge in generating appropriate code for `pulumi import` lies in
+determining exactly what the input values for a particular resource should be.
+In many providers, it is not necessarily possible to accurately recover a
+resource's inputs from its state. This observation led to the diff-oriented
+approach described above, where the importer begins with an extremely minimal
+set of inputs and attempts to derive the actual inputs from the results of a
+call to the provider's [](pulumirpc.ResourceProvider.Diff) method.
+Unfortunately, the results are not always satisfactory, and the relatively small
+set of inputs present in the generated code can make it difficult for users to
+determine what inputs they *actually* need to pass to the resource to describe
+its current state.
+
+A few other approaches might be:
```

Comment:
Should the other approaches move out of the docs and into tickets /RFCs ? It's hard to mix "what is" and "what might be" in the same doc? YMMV though.

Reply:
As above, I've just copied this from what's there, but I agree there's an opportunity to clean this up. I've made a note and will sort in a future PR.